bin/upgrade.py

```diff
@@ -116,14 +119,50 @@ def rename_testcase(old_base: Path, new_dir: Path) -> None:
     if bad_dir.is_dir() and not any(bad_dir.iterdir()):
         bad_dir.rmdir()
 
+    # Move .hint and .desc files to the Test Case Configuration .yaml file
+
+    test_case_yamls = defaultdict[Path, CommentedMap](CommentedMap)
+    for f in (problem_path / "data").rglob("*.yaml"):
+        if f.with_suffix(".in").exists():  # Prevent reading test_group.yaml, which has no *.in file
```

Comment:
does the `generato.yaml` prohibit the unnumbered test case name `test_group`?

Reply:
One more remark, do we also want to check this for not generated testcases and if yes where?
- during `bt validate`?
- in `problem.testcases()` ? (would probably need caching to not get the same message multiple times?)

Reply:
Good call :smile: I put the checks in `problem.testcases()`, because I think this should always be checked, and it feels like the place to do it: we also check for `.in`/`.ans` consistency there.

Regarding caching: Placing the checks in `Problem.validate_data` would also show the warning multiple times, since this function is called once for each `mode`. I think it's fine if the warning is shown multiple times? It should not occur that often anyway, and I don't immediately see a clean way of implementing caching without writing things to the tmpdir (we could store the warnings in some global variable, but that feels ugly) :stuck_out_tongue:

Reply:
we already have `problem._testcases` cache we could add something like `problem._warned_test_group` (or some nicer name) and just warn for a testcase if it is not yet in that set.

I thinks its fine to warn for `sample/test_group.in` and `secret/test_group.in` individually, but its a bit weird if we warn for the exact same file during input, answer and output validation and run and...

Reply:
Ah thanks, that's already better than a global cache :smile: I think I've made it look somewhat clean, what do you think? :slightly_smiling_face: